sdk/include/ddk/ntstrsafe.h

```diff
@@ -2243,350 +2486,441 @@ NTSTRSAFEAPI RtlStringCatExWorkerW(STRSAFE_LPWSTR pszDest,size_t cchDest,size_t
                 }
             }
+
+            if (!pszSrc)
+                pszSrc = "";
+        }
+        else
+        {
+            Status = RtlStringLengthWorkerA(pszDest, cchDest, &cchDestLength);
             if (NT_SUCCESS(Status))
             {
-                if (cchDest==0)
+                pszDestEnd = pszDest + cchDestLength;
+                cchRemaining = cchDest - cchDestLength;
+            }
+        }
+
+        if (NT_SUCCESS(Status))
+        {
+            if (cchDest == 0)
+            {
+                if ((cchToAppend != 0) && (*pszSrc != '\0'))
                 {
-                    if ((cchToAppend!=0) && (*pszSrc!='\0'))
-                    {
-                        if (!pszDest)
-                            Status = STATUS_INVALID_PARAMETER;
-                        else
-                            Status = STATUS_BUFFER_OVERFLOW;
-                    }
+                    if (!pszDest)
+                        Status = STATUS_INVALID_PARAMETER;
+                    else
+                        Status = STATUS_BUFFER_OVERFLOW;
                 }
-                else
-                    Status = RtlStringCopyNExWorkerA(pszDestEnd,cchRemaining,(cchRemaining*sizeof(char)) + (cbDest % sizeof(char)),pszSrc,cchToAppend,&pszDestEnd,&cchRemaining,dwFlags & (~(STRSAFE_FILL_ON_FAILURE | STRSAFE_NULL_ON_FAILURE)));
+            }
+            else
+            {
+                Status = RtlStringCopyNExWorkerA(pszDestEnd, cchRemaining, (cchRemaining * sizeof(char)) + (cbDest % sizeof(char)), pszSrc, cchToAppend, &pszDestEnd, &cchRemaining, dwFlags & (~(STRSAFE_FILL_ON_FAILURE | STRSAFE_NULL_ON_FAILURE)));
             }
         }
+    }
+
     if (!NT_SUCCESS(Status))
     {
         if (pszDest)
         {
             if (dwFlags & STRSAFE_FILL_ON_FAILURE)
             {
-                memset(pszDest,STRSAFE_GET_FILL_PATTERN(dwFlags),cbDest);
-                if (STRSAFE_GET_FILL_PATTERN(dwFlags)==0)
+                memset(pszDest, STRSAFE_GET_FILL_PATTERN(dwFlags), cbDest);
+                if (STRSAFE_GET_FILL_PATTERN(dwFlags) == 0)
                 {
                     pszDestEnd = pszDest;
                     cchRemaining = cchDest;
                 }
-                else
-                    if (cchDest > 0)
-                    {
-                        pszDestEnd = pszDest + cchDest - 1;
-                        cchRemaining = 1;
-                        *pszDestEnd = '\0';
-                    }
-            }
-            if (dwFlags & (STRSAFE_NULL_ON_FAILURE))
-            {
-                if (cchDest > 0)
+                else if (cchDest > 0)
                 {
-                    pszDestEnd = pszDest;
-                    cchRemaining = cchDest;
+                    pszDestEnd = pszDest + cchDest - 1;
+                    cchRemaining = 1;
                     *pszDestEnd = '\0';
                 }
             }
+
+            if (dwFlags & (STRSAFE_NULL_ON_FAILURE) && cchDest > 0)
```

Comment:
Ditto. But the parentheses around the single STRSAFE_NULL_ON_FAILURE name are unneeded.